### PR TITLE
linux_xanmod: match features on homepage

### PR DESCRIPTION
**Pull request by moni-dz**
###### Motivation for this change
Looking at the kernel config, some things that xanmod users would expect aren't present, such as having `PREEMPT` enabled.

This PR aims to match the claims of the website, specifically:

- [x] Enabling preemption (`PREEMPT`)
- [x] Allow kernel to be tickless when possible (`NO_HZ_FULL`)
- [x] Have a 500 Hz tick rate (`HZ_500`)

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by fortuneteller2k** [comment]
successfully booted, `PREEMPT` is enabled
```
~ took 15s ^ uname -a
Linux superfluous 5.12.13-xanmod1-cacule #1-NixOS SMP PREEMPT Tue Jan 1 00:00:00 UTC 1980 x86_64 GNU/Linux
```

**Review by fufexan — APPROVED** [review]

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 128007` at 167fde46 run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>42 packages failed to build:</summary>
  <ul>
    <li><a href="https://gist.github.com/8c3d4f3a334740f7dd44a532baeaef93">linuxPackages_xanmod.acpi_call</a></li>
    <li><a href="https://gist.github.com/91b911274a1a9e38fa6e3c58eb6c56f0">linuxPackages_xanmod.akvcam</a></li>
    <li><a href="https://gist.github.com/dbb7f1b9e88acd8f310c4ebebe676132">linuxPackages_xanmod.anbox</a></li>
    <li><a href="https://gist.github.com/b0b21d9279cafdcc2fca057322e59fe6">linuxPackages_xanmod.batman_adv</a></li>
    <li><a href="https://gist.github.com/390127a110134ec7cc24397ac83902c3">linuxPackages_xanmod.bcc</a></li>
    <li><a href="https://gist.github.com/c0d2845aeb112f6b7caf73d9a6515486">linuxPackages_xanmod.can-isotp</a></li>
    <li><a href="https://gist.github.com/983783ce8d5f0f5716edcb4501c84d62">linuxPackages_xanmod.cryptodev</a></li>
    <li><a href="https://gist.github.com/b1b3aa4a42df311fb4b9cdc3c7fcde0f">linuxPackages_xanmod.ddcci-driver</a></li>
    <li><a href="https://gist.github.com/099944500f960494c2df2dca3f8b30cb">linuxPackages_xanmod.digimend</a></li>
    <li><a href="https://gist.github.com/c9acb80dbec7a1c26bec298bc8829344">linuxPackages_xanmod.ena</a></li>
    <li><a href="https://gist.github.com/f1c5a491b3616ba579f0654bf71c60b1">linuxPackages_xanmod.evdi</a></li>
    <li><a href="https://gist.github.com/53dad9773bb57930ab9c13dba3a4e162">linuxPackages_xanmod.fwts-efi-runtime</a></li>
    <li><a href="https://gist.github.com/a09fca9a39a2b9645e2b5a9968e3a887">linuxPackages_xanmod.gcadapter-oc-kmod</a></li>
    <li><a href="https://gist.github.com/5b4f40cef7ffdffcd5dc8084c9b994a5">linuxPackages_xanmod.hid-nintendo</a></li>
    <li><a href="https://gist.github.com/a3def36c6300daa5ff7a66abea74094a">linuxPackages_xanmod.jool</a></li>
    <li><a href="https://gist.github.com/7b55cd5206446cfefbf220f6faf9e04e">linux_xanmod</a> (linuxPackages_xanmod.kernel)</li>
    <li><a href="https://gist.github.com/8906a36c892a19408e47b03872c73904">linuxPackages_xanmod.lttng-modules</a></li>
    <li><a href="https://gist.github.com/f61d10d956a1c2ac8b6cb289c71be8bf">linuxPackages_xanmod.mba6x_bl</a></li>
    <li><a href="https://gist.github.com/f68ffff7c00f909d05a71746c20cac42">linuxPackages_xanmod.mbp2018-bridge-drv</a></li>
    <li><a href="https://gist.github.com/36c961b1bebaa682ea6c2922ade047e9">linuxPackages_xanmod.mwprocapture</a></li>
    <li><a href="https://gist.github.com/17a4e32eed8a4d27d32d3276d20c4159">linuxPackages_xanmod.mxu11x0</a></li>
    <li><a href="https://gist.github.com/b1e32b1ed20d4f97603685c199ccaf6d">linuxPackages_xanmod.netatop</a></li>
    <li><a href="https://gist.github.com/672c5cc9126d21633dcd5566c986865a">linuxPackages_xanmod.oci-seccomp-bpf-hook</a></li>
    <li><a href="https://gist.github.com/8b67374d3dc729e605ebe8bb86bfd88d">linuxPackages_xanmod.openrazer</a></li>
    <li><a href="https://gist.github.com/01e17091ce5406dea890d87bdce2b0c5">linuxPackages_xanmod.r8125</a></li>
    <li><a href="https://gist.github.com/f94d0cf14a7301887273d0cb8c1bb623">linuxPackages_xanmod.r8168</a></li>
    <li><a href="https://gist.github.com/31dba5029d476f5b1b0252a3fe90666f">linuxPackages_xanmod.rtl8192eu</a></li>
    <li><a href="https://gist.github.com/4a6f8e96e7d47ba701bdedc2ade1486e">linuxPackages_xanmod.rtl8723bs</a></li>
    <li><a href="https://gist.github.com/7497b1c0cec35036b6fbdedf6843ba3d">linuxPackages_xanmod.rtl8812au</a></li>
    <li><a href="https://gist.github.com/19d4bd4f297ca68b319e648a06bb6be4">linuxPackages_xanmod.rtl8821ce</a></li>
    <li><a href="https://gist.github.com/e6fea800bb08bf374a5ce604099077bf">linuxPackages_xanmod.rtl8821cu</a></li>
    <li><a href="https://gist.github.com/0ae651c94256c72363f1f4036c304613">linuxPackages_xanmod.rtl88x2bu</a></li>
    <li><a href="https://gist.github.com/3f5d59718705fbd0bd894849ebacac7d">linuxPackages_xanmod.rtw88</a> (linuxPackages_xanmod.rtlwifi_new)</li>
    <li><a href="https://gist.github.com/db53b48d41e6d7a9cf4f37b9b5aadfca">linuxPackages_xanmod.sch_cake</a></li>
    <li><a href="https://gist.github.com/7a382cc10e6725ede189a37d682cf158">linuxPackages_xanmod.systemtap</a></li>
    <li><a href="https://gist.github.com/99e32710e8435ecde9e3e582b9985578">linuxPackages_xanmod.tuxedo-keyboard</a></li>
    <li><a href="https://gist.github.com/a0f646c68fb5080b5b53fb9cd93ab4cb">linuxPackages_xanmod.v4l2loopback</a></li>
    <li><a href="https://gist.github.com/6f185ae69880962cd3ec66e4eb0cf1e9">linuxPackages_xanmod.veikk-linux-driver</a></li>
    <li><a href="https://gist.github.com/7a213227026afb78dcd1978dce8c44f7">linuxPackages_xanmod.vhba</a></li>
    <li><a href="https://gist.github.com/17bab15ce22e4c23db1fb136368e8077">linuxPackages_xanmod.xpadneo</a></li>
    <li><a href="https://gist.github.com/5e00d09b334663ca73c78f6c93b6bde4">linuxPackages_xanmod.zfs</a> (linuxPackages_xanmod.zfsStable)</li>
    <li><a href="https://gist.github.com/0eb8b82789cfbf115696109790412d4a">linuxPackages_xanmod.zfsUnstable</a></li>
  </ul>
</details>
<details>
  <summary>3 packages skipped due to time constraints:</summary>
  <ul>
    <li>linuxPackages_xanmod.dpdk-kmods</li>
    <li>linuxPackages_xanmod.openafs (linuxPackages_xanmod.openafs_1_8)</li>
    <li>linuxPackages_xanmod.openafs_1_9</li>
  </ul>
</details>

Note that build failures [may predate](https://github.com/nix-community/hydra-check) this PR, and could be nondeterministic or hardware dependent.
Please exercise your independent judgement. Does something look off? Please file an [issue](https://github.com/rmcgibbo/r-rmcgibbo/issues) or reach out on IRC.

**Review by samuelgrf — APPROVED** [review]
Successfully built and booted

**Comment by fortuneteller2k** [comment]
Website also describes the kernel to be tickless and to be at 500 Hz, so I added both of those.

Successfully booted with `PREEMPT` enabled, `NO_HZ_FULL` enabled, and `HZ` set to 500.

```
nixpkgs on  preempt-xanmod ^ uname -a
Linux superfluous 5.12.13-xanmod1-cacule #1-NixOS SMP PREEMPT Tue Jan 1 00:00:00 UTC 1980 x86_64 GNU/Linux

nixpkgs on  preempt-xanmod ^ zcat /proc/config.gz | rg HZ_500
CONFIG_HZ_500=y

nixpkgs on  preempt-xanmod ^ zcat /proc/config.gz | rg NO_HZ_FULL
CONFIG_NO_HZ_FULL=y
```

File: pkgs/os-specific/linux/kernel/linux-xanmod.nix [hunk 1]
```diff
@@ -15,6 +15,13 @@ buildLinux (args // rec {
     sha256 = "sha256-eFIWlguU1hnkAgTbRxSMTStq0X7XW4IT1/9XlQSgdMQ=";
   };
 
+  structuredExtraConfig = with lib.kernel; {
+    PREEMPT = lib.mkForce yes;
```
Comment (samuelgrf) [inline pkgs/os-specific/linux/kernel/linux-xanmod.nix:19]: Additional changes, so all the features mentioned on the homepage are applied (WIP untested):

```suggestion
  
    # Preemptive Full Tickless Kernel at 500Hz.
    PREEMPT = lib.mkForce yes;
    PREEMPT_VOLUNTARY = lib.mkForce no;
    NO_HZ_FULL = yes;
    HZ_500 = yes;

    # Google's Multigenerational LRU Framework.
    LRU_GEN = yes;

    # Google's BBRv2 TCP congestion Control.
    DEFAULT_BBR2 = yes;
    DEFAULT_TCP_CONG = freeform "bbr2";
    TCP_CONG_BBR2 = yes;

    # FQ-PIE Packet Scheduling.
    NET_SCH_DEFAULT = yes;
    DEFAULT_FQ_PIE = yes;
    DEFAULT_NET_SCH = freeform "fq_pie";

    # Graysky's additional CPU optimizations for GCC / Clang.
    CC_OPTIMIZE_FOR_PERFORMANCE_O3 = yes;

    # Android Ashmem and Binder IPC Driver as module for Anbox.
    ANDROID = yes;
    ANDROID_BINDERFS = module;
    ANDROID_BINDER_DEVICES = freeform "binder,hwbinder,vndbinder";
    ANDROID_BINDER_IPC = module;
    ASHMEM = module;
```

Comment (samuelgrf) [reply pkgs/os-specific/linux/kernel/linux-xanmod.nix:19]: I submitted the above review by accident, this PR can be merged as is.

I'll open a seperate PR once I have built and tested the kernel with these changes.

Comment (fortuneteller2k) [reply pkgs/os-specific/linux/kernel/linux-xanmod.nix:19]: Would appreciate it, thanks!

**Pull request merged**